foxpuppet/windows/browser.py

```diff
@@ -2,43 +2,59 @@
 # License, v. 2.0. If a copy of the MPL was not distributed with this file,
 # You can obtain one at http://mozilla.org/MPL/2.0/.
 
-from foxpuppet.windows import Windows
-
 from selenium.webdriver.common.by import By
 
+from foxpuppet.expected import new_browser_window_is_opened
+from foxpuppet.windows.base import BaseWindow
+
+
+class BrowserWindow(BaseWindow):
+
+    """Representation of a browser window."""
 
-class BrowserWindow(object):
     _file_menu_button_locator = (By.ID, 'file-menu')
     _file_menu_private_window_locator = (By.ID, 'menu_newPrivateWindow')
     _file_menu_new_window_button_locator = (By.ID, 'menu_newNavigator')
     _nav_bar_locator = (By.ID, 'nav-bar')
     _tab_browser_locator = (By.ID, 'tabbrowser-tabs')
 
-    def __init__(self, selenium, *args, **kwargs):
-        self.selenium = selenium
-        self._windows = Windows(selenium)
+    def __init__(self, *args, **kwargs):
+        super(BrowserWindow, self).__init__(*args, **kwargs)
 
     @property
     def is_private(self):
         """Returns True if this is a Private Browsing window."""
 
-        self.selenium.set_context('chrome')
-        return self.selenium.execute_script("""
+        self.selenium.switch_to.window(self.handle)
+        with self.selenium.context('chrome'):
+            return self.selenium.execute_script(
+                """
                 Components.utils.import("resource://gre/modules/PrivateBrowsingUtils.jsm");
 
                 let chromeWindow = arguments[0].ownerDocument.defaultView;
                 return PrivateBrowsingUtils.isWindowPrivate(chromeWindow);
-            """, self._windows.window_element)
-        self.selenium.set_context('content')
+                """, self.document_element)
 
     def open_window(self, private=False):
-        self.selenium.set_context('chrome')
-        self.selenium.find_element(*self._file_menu_button_locator).click()
-        with self._windows.wait_for_new_window():
+        """Opens a new browser window
+
+        :param private: Optional parameter to open a private browsing window.
+                        Defaults to false.
+        """
+
+        handles_before = self.selenium.window_handles
+        self.selenium.switch_to.window(self.handle)
+
+        with self.selenium.context('chrome'):
+            # Opens private or non-private window
+            self.selenium.find_element(*self._file_menu_button_locator).click()
             if private:
                 self.selenium.find_element(
                     *self._file_menu_private_window_locator).click()
             else:
                 self.selenium.find_element(
                     *self._file_menu_new_window_button_locator).click()
-        self.selenium.set_context('content')
+
+        return self.wait.until(
+            new_browser_window_is_opened(self.selenium, handles_before),
+            message="No new browser window opened")
```

Comment:
As mentioned earlier, I'm not a bit fan of this logic. The handling of window handles should be transparent to the caller. So context manager or a callback sounds better alternatives.